### PR TITLE
feat(osv): read npm dependencies from package.json

### LINKED ISSUE #227: [FEATURE] Support package.json manifests by OSV Analyzer
## Feature Description

Let the OSV Analyzer read `package.json`, so npm dependencies are checked for known vulnerabilities the same way Python ones are.

  ## Problem Statement

Many of the skills I scan are JavaScript and list their dependencies in `package.json` so I'd like to cover it with skill-scanner.

Right now only Python manifest files are read: `requirements.txt`, `pyproject.toml`, `setup.cfg`, `setup.py` and `Pipfile`.

  ## Use Case

  Scanning JavaScript skills that declare npm dependencies in `package.json`. No new flag is needed: `--use-osv` and the default static analyzer cover it.

  ## Benefits

  - Closes a gap that is invisible today.

  ## Checklist

  - [x] I have searched existing issues/PRs for similar requests
  - [x] I have described a clear use case
  - [x] I have considered implementation complexity
  - [x] This aligns with the project's goals (security scanning for Agent Skills)

### PR DESCRIPTION
**Pull request by chinskiy**
# Pull Request

## Description

The OSV analyzer and the unpinned-dependency check read Python manifests only, so
a skill declaring npm dependencies is scanned as if it declared none — a clean
report because nothing was looked up.

This reads `package.json` for both: `--use-osv` queries exact pins against OSV's
`npm` ecosystem, and the static check flags ranges like `^4.17.0`. Manifests only,
no lockfiles, matching how Python dependencies are collected. No new
dependencies, flags or rule IDs.

## Type of Change

- [x] New feature (non-breaking change that adds functionality)

## Related Issues

Closes #227

## Changes Made

- **`npm_manifest.py` (new)** — manifest parsing for both checks. Reads
  `dependencies`, `devDependencies` and `optionalDependencies`; skips
  `peerDependencies`, `bundleDependencies`, artifact references (`file:`, `git+`,
  `github:`, tarballs) and anything under `node_modules/`. Exact means `1.2.3`,
  `=1.2.3`, `v1.2.3` or `npm:real-pkg@1.2.3`, resolved to the real package name.
- **OSV analyzer** — `querybatch` already takes an ecosystem per query, so the
  internal record was widened and Python and npm go out together. Adds dedup on
  `(ecosystem, name, version)`, request chunking, a per-skill cap, and per-chunk
  fail-open.
- **Static analyzer** — classifies npm specs (`^`/`~`/`>=`/`*`/tags → unpinned
  MEDIUM; `1.x`/`1.2.*` → wildcard LOW) and **scopes lockfile suppression per
  ecosystem**, so a `yarn.lock` can't silence Python findings. PyPI behaviour is
  unchanged, since JS lockfile names weren't in the old set.
- **Docs** — npm source and exclusions, the new bounds, and an explicit note that
  no lockfile is read in either ecosystem.

### Test Coverage

- [x] Unit tests added/updated
- [x] Integration tests added/updated
- [x] All tests pass locally
- [x] Test coverage maintained or improved

### Manual Testing

```bash
skill-scanner scan ./demo-skill --use-osv
```

**Results:**
- Expected: exact pin looked up, range flagged, peer dep ignored
- Actual:
  ```
  [HIGH  ] SUPPLY_CHAIN_KNOWN_VULNERABILITY  package.json:5  lodash@4.17.15
  [MEDIUM] SUPPLY_CHAIN_UNPINNED_DEPENDENCY  package.json:6  "chalk": "^5.0.0"
  ```

Also run on real repositories: one had 16 exact pins and both of the two carrying
advisories were reported (`vitest@3.2.7`, `yaml@2.8.1`). Another had a
`package-lock.json` at root plus a Python range in a skill — the Python finding is
still reported, which is what the per-ecosystem scoping preserves.

### Code Quality
- [x] Code follows project style guidelines
- [x] Type hints added where applicable
- [x] Docstrings added/updated for public APIs
- [x] No hardcoded credentials or secrets
- [x] Error handling is comprehensive
- [x] Logging is appropriate

### Documentation
- [x] README updated (if needed) — none required
- [x] API documentation updated (if needed)
- [ ] CHANGELOG updated — no CHANGELOG file in this repository
- [x] Code comments added for complex logic

### Security
- [x] No new security vulnerabilities introduced
- [x] Input validation added where needed
- [x] Follows security best practices from workspace rules
- [x] No eval/exec on user input without sanitization

### Testing
- [x] Tests pass: `uv run pre-commit run --all-files` — 16/16 hooks
- [x] Benchmark passes: `uv run python evals/runners/benchmark_runner.py` — 24 skills, 0 errors, 100% accuracy and precision
- [x] No regressions in existing functionality
- [x] Edge cases covered

Also ran `validate-rules` (46 signature, 927 manifest rules, cel-go type-check OK)
and `pytest tests/test_rule_registry.py tests/test_static_policy_integration.py`.

## Performance Impact

- [x] No significant performance regression
- [x] Performance benchmarks run (if applicable)
- [x] Resource usage is acceptable

## Additional Notes

1. **Transitive dependencies are out of scope** — only what the manifest declares
   is seen. Lockfiles would cover the rest but break the symmetry with Python and
   bring an advisory-volume problem. Happy to open that separately.
2. **`chunk_size` and `max_dependencies` are constructor arguments, not policy
   knobs**, matching the analyzer's current style. Happy to wire them into
   `scan_policy.py` if preferred.
3. **`docs-site` still lists Ruby/Go/Rust/Maven/NuGet/Composer/Swift** for
   unpinned dependencies. Left untouched as it predates this change, though
   `static-analyzer.md` (updated here) now lists only Python plus `package.json`.

Three commits, each independently valid: OSV half, static half, docs.

## Summary by CodeRabbit

- **New Features**
  - OSV vulnerability scanning now supports exactly pinned JavaScript dependencies in `package.json`, alongside Python dependencies.
  - JavaScript dependency declarations are checked for exact pinning, including runtime, development, optional, and aliased packages.
  - Vulnerability lookups are deduplicated and processed in bounded batches, with failures isolated to individual batches.

- **Documentation**
  - Updated analyzer, CLI, architecture, FAQ, and feature documentation to describe JavaScript coverage, pinning rules, lockfile limitations, and dependency handling.

### CODE REVIEW

File: skill_scanner/core/analyzers/static.py [hunk 1]
```diff
@@ -2503,60 +2566,53 @@ def _check_dependency_pinning(self, skill: Skill) -> list[Finding]:
...
+        Sources checked -- PyPI: ``requirements*.txt``, ``pyproject.toml``
         (``[project]`` dependencies and optional-dependencies), ``setup.cfg``,
         ``setup.py`` (``install_requires``), ``Pipfile``, and a
-        ``dependencies`` list under manifest ``metadata``.
+        ``dependencies`` list under manifest ``metadata``.  npm:
+        ``package.json`` (``dependencies``, ``devDependencies`` and
+        ``optionalDependencies``).
         """
         findings: list[Finding] = []
+        present = {Path(f.relative_path).name.lower() for f in skill.files}
```
[the remark this answers is not in the document]
Comment (chinskiy) [reply skill_scanner/core/analyzers/static.py]: Fixed in 8ef66e6.

Confirmed before fixing: a skill with its own `package.json` plus a vendored `node_modules/vendor/yarn.lock` went from `[requirements.txt, package.json]` to `[requirements.txt]` — the own-manifest finding disappeared.

This was an inconsistency on my side rather than an edge case: `is_npm_manifest()` already excluded vendored *manifests* on the grounds that `node_modules/` content is not the author's, but I computed the lockfile-presence set over every file. Same rule, applied to one half.

Extracted the path check as `is_vendored()` and used it in both places, so the two cannot drift again. Regression test: `test_vendored_lockfile_does_not_suppress_own_manifest`.

File: skill_scanner/core/analyzers/osv_analyzer.py [hunk 2]
```diff
@@ -258,51 +327,73 @@ def _parse_pinned(raw: str) -> tuple[str, str] | None:
...
+                for dependency in dependencies
             ]
         }
         response = self._client.post(self.QUERYBATCH_URL, json=payload)
         response.raise_for_status()
         results = response.json().get("results", [])
-        return [(results[index].get("vulns") or []) if index < len(results) else [] for index in range(len(packages))]
+        return [
+            (results[index].get("vulns") or []) if index < len(results) else [] for index in range(len(dependencies))
+        ]
```
[the remark this answers is not in the document]
Comment (chinskiy) [reply skill_scanner/core/analyzers/osv_analyzer.py]: Fixed in 8ef66e6.

Reproduced both paths: `{"results": ["nope"]}` raised `AttributeError: 'str' object has no attribute 'get'`, and a non-mapping entry inside `vulns` failed the same way in `_create_finding()`. Neither is caught by `(httpx.HTTPError, ValueError)`, so the fail-open behaviour the module docstring promises did not actually hold for a malformed body.

Added `_vulns_per_dependency()`, which validates that the body is an object and `results` is a list, filters `vulns` to mappings, and raises `ValueError` on a bad shape — the per-chunk handler already treats that as a failed chunk, so this now degrades that chunk's coverage instead of aborting the scan.

Tests: `test_non_mapping_result_entry_fails_open`, `test_results_not_a_list_fails_open`, and `test_non_mapping_vulnerability_entry_is_ignored` (which asserts the valid advisory in a mixed list is still reported).

File: skill_scanner/core/analyzers/npm_manifest.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        return []
+    if not isinstance(data, dict):
+        return []
+
+    entries: list[DeclaredDependency] = []
+    for section, is_dev in _DEPENDENCY_SECTIONS:
+        declared = data.get(section)
+        if not isinstance(declared, dict):
+            continue
+        for key, spec in declared.items():
```
[the remark this answers is not in the document]
Comment (chinskiy) [reply skill_scanner/core/analyzers/npm_manifest.py]: Fixed in 8ef66e6.

Verified the npm behaviour against the docs line you cited, and reproduced the parser output: a name in both sections emitted `[('foo', '1.0.0'), ('foo', '2.0.0')]`, so the advisory lookup could query `1.0.0` — a version npm never installs — and the pinning check could report the overridden range.

`optionalDependencies` names are now collected first and skipped while walking `dependencies`. Matching is on the manifest key, since that is what npm dedupes on, so aliased entries resolve after the override is applied. `devDependencies` is left alone: npm treats a name in both `dependencies` and `devDependencies` as an error rather than an override, so there is no defined precedence to implement there.

Test: `test_optional_dependency_overrides_duplicate_name`.

File: skill_scanner/core/analyzers/osv_analyzer.py [hunk 4]
```diff
@@ -159,21 +178,31 @@ def _entries_from_pipfile(path: str, content: str) -> list[tuple[str, int | None
...
         policy: ScanPolicy | None = None,
+        chunk_size: int | None = None,
+        max_dependencies: int | None = None,
     ):
         super().__init__("osv_analyzer", policy=policy)
         self.enabled = enabled
         self.ecosystem = ecosystem
         self.timeout = timeout
+        self.chunk_size = chunk_size or self.QUERY_CHUNK_SIZE
+        self.max_dependencies = max_dependencies or self.MAX_DEPENDENCIES
```
[the remark this answers is not in the document]
Comment (chinskiy) [reply skill_scanner/core/analyzers/osv_analyzer.py]: Fixed in 8ef66e6.

I was initially inclined to leave this — neither bound is reachable from the CLI or the API, so no user input flows in. Reproducing it changed my mind: the failure mode is silence rather than an exception. `chunk_size=-5` returned zero query chunks, so every finding vanished with nothing logged, and `max_dependencies=-1` was stored as-is, making `del deduplicated[-1:]` drop the last entry instead of capping.

`_positive_bound()` now rejects zero, negatives and non-integers (bools included, since `bool` is an `int` subclass), and keeps `None` as the default sentinel.

Tests: `TestBoundsValidation`, parametrised over `0`, `-1`, `-5`, `1.5` and `"5"`, plus a case asserting `None` still yields the class defaults.
